user_groups: Show details in group deactivation error.

sahil839 opened the pull request:
<!-- Describe your pull request here.-->
- First commit adds backend code to include the details of supergroups and of the groups and streams for which the group is being used as setting.
- Second commit adds the frontend code to show the details.

Fixes:  <!-- Issue link, or clear description.--> https://chat.zulip.org/#narrow/stream/101-design/topic/Support.20user.20group.20deactivation/near/1949585

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
![image](https://github.com/user-attachments/assets/33e367f7-845a-4382-87c0-13214bc8a14a)

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

sahil839 commented on zerver/lib/exceptions.py:
```diff
@@ -715,3 +716,19 @@ def __init__(self, first_parameter: str, second_parameter: str) -> None:
+
+
+class CannotDeactivateGroupInUseError(JsonableError):
+    code = ErrorCode.CANNOT_DEACTIVATE_GROUP_IN_USE
```
Using a different code, so that client can show error accordingly.

timabbott replied:
Yes, sounds good.

sahil839 commented on zerver/lib/user_groups.py:
```diff
@@ -185,35 +186,51 @@ def access_user_group_for_deactivation(
+            in setting_group_ids_using_deactivating_user_group
+        ]
+        if len(objection_settings) > 0:
+            objections.append(dict(type="stream", stream_id=stream.id, settings=objection_settings))
```
Do we want to use `channel` in type here? I see we still use `stream` for variables.

timabbott replied:
I think when adding something new like this to the API (in this case, an error response format), we decided that using `channel` is desirable. We've not attempted to bulk-migrate the server to use `channel`.

sahil839 replied:
Updated the type to `channel` but the variables, both in server and webapp, still use stream.

sahil839 commented:
We do not handle inaccessible streams right now. Might work fine for now since there is only one group setting for stream and that cannot be set to user defined groups.

timabbott commented on zerver/lib/user_groups.py:
```diff
@@ -185,35 +186,51 @@ def access_user_group_for_deactivation(
-        .filter(stream_setting_query)
-        .exists()
+    for stream in Stream.objects.filter(realm_id=user_group.realm_id, deactivated=False).filter(
+        stream_setting_query
```
`setting_group_ids_using_deactivating_user_group` likely needs to be converted to a `Set` for performance reasons.

sahil839 replied:
Updated.

timabbott commented on web/templates/user_group_settings/cannot_deactivate_group_banner.hbs:
```diff
@@ -0,0 +1,14 @@
+            <li><a href="#organization">Organization settings</a></li>
+        {{/if}}
+    </ul>
+</p>
```
This is missing translation tags.

sahil839 replied:
My bad. Added.

timabbott commented:
> We do not handle inaccessible streams right now. Might work fine for now since there is only one group setting for stream and that cannot be set to user defined groups.

Yeah, this is fine. I don't think we need to track this for now; we'll hear about it if it creates a problem down the line.)

timabbott commented:
Looks great, posted a few comments but this should be good to merge after addressing them. Please add the integration review label when ready.

sahil839 commented:
Updated API docs and added changelog entry.

Also added commit to handle inaccessible streams.

![Screenshot from 2024-10-01 15-22-13](https://github.com/user-attachments/assets/9dc84207-df03-4dbb-8fd7-f0fb7844c327)

timabbott commented:
Oh, I had the thought that we could make the "Unknown stream" entries still links, just with `{channel_id}-{unknown}` as the relevant URL component.

Also that should be "Unknown channel"; "stream" isn't supposed to be in user-facing strings.

sahil839 commented:
Not sure if linking the "Unknown channel" text would be helpful. It will just take to the "subscribed" list in streams overlay with right panel being empty.

timabbott commented:
Well it'd be helpful in that you can share the link with an administrator who could help resolve the issue. But this isn't very likely to come up; most people deactivating groups are admins, and they can see every channel.

alya commented:
@timabbott points out that the link might be helpful to share with someone else. Not super important, but I guess we can add it.
